### ISSUE TITLE
Quad9: failing to encode authority for some queries.

### ISSUE THREAD

**Issue by bmrzycki** [aarond10/https_dns_proxy#50]
I am seeing a an issue when using [DoH on quad9](https://quad9.net/doh-quad9-dns-servers/). I have tried using all three of their configuration options and see similar warning strings.

I should note this doesn't happen for all host queries, only some of them. For example, `yahoo.com` **does not** produce the warnings.

I'll also note this configuration is working with `dnsmasq` to resolve hosts. It's just that `https_dns_proxy` reports some of these packets as malformed, which `dnsmasq` seems to ignore with my configuration.

Here are the steps to reproduce the warnings:

1. execution:
```
./https_dns_proxy \
    -u nobody -g nogroup \
    -b 9.9.9.9,149.112.112.9 \
    -r https://dns9.quad9.net/dns-query?ct=application/dns-json& \
    -a 127.0.0.1 -p 5054 \
    -d -v -l /root/gw/https_dns_proxy/run_quad9.log
```

2. local query:
```
dig -p 5054 @127.0.0.1 tensorflow.com
```

3. log result:
```
[W] 1548086891.902399 json_to_dns.c:416 Failed to encode Authority ix 1.
[W] 1548086891.902453 json_to_dns.c:416 Failed to encode Authority ix 2.
[W] 1548086891.902462 json_to_dns.c:416 Failed to encode Authority ix 3.
[W] 1548086891.902469 json_to_dns.c:416 Failed to encode Authority ix 4.
[W] 1548086891.902477 json_to_dns.c:416 Failed to encode Authority ix 5.
[W] 1548086891.902483 json_to_dns.c:416 Failed to encode Authority ix 6.
[W] 1548086891.902490 json_to_dns.c:416 Failed to encode Authority ix 7.
```

**Comment by aarond10** [comment]
These look like DNSSEC related responses that aren't supported yet (RRSIG, NSEC3):
```
$ dig tensorflow.com  +dnssec
;; AUTHORITY SECTION:
com.                    900     IN      SOA     a.gtld-servers.net. nstld.verisign-grs.com. 1548326650 1800 900 604800 86400
com.                    900     IN      RRSIG   SOA 8 1 900 20190131104410 20190124093410 16883 com. R+n56SAVIUFdXWKWN6E/jchYEYCgLqC8kQ4xSTgzSGKk+HYC0L8Bnnw1 Fowakx7TV0ft+FjFQdjv+CRYfn3Vuwknkjz7TvFXOftMV+nX20y9Uxsr oEeEVhLp6YcmliZdoGeXEcamw+ukfzXzHVeeV5rIPsjAmnW0x+iGmLxG Ebs=
ck0pojmg874ljref7efn8430qvit8bsm.com. 86400 IN NSEC3 1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM
ck0pojmg874ljref7efn8430qvit8bsm.com. 86400 IN RRSIG NSEC3 8 2 86400 20190131054530 20190124043530 16883 com. eg1n7WXtR3qpZdjJTrXm3VEqhRWwaMXbomHWLWO3+Mixz/f/h6DBNEre p2ua4BLyXKu6ZjFFeqpMWacSSLUULHt7U50ag5dJJm6EaSRunqsddPSX GLpVrikC+Xh4Y1de3bMzgz//7+/+GgNLHeGr+ko+KcS2D//B5sz+LmEb oqw=
ek3b4b8dsur37cd5jut6ruld7ps3b17o.com. 86400 IN NSEC3 1 1 0 - EK3C80TGFMR3PIS2U1G8SH6PQLT1I6LT NS DS RRSIG
ek3b4b8dsur37cd5jut6ruld7ps3b17o.com. 86400 IN RRSIG NSEC3 8 2 86400 20190129053228 20190122042228 16883 com. DHp6ml2ufVEB39TRdKByFVHk0Mw2U+ZazuTWedanqfTEk7zwGLRAwCOn eCa3RF7SDXI5rXXMz8CYUwpPLAPLQGG36CshNVHyEytseBK6pnPEzs1x rlklNtRVJ+dKszxwJYJfhieSXA0S4fi27JYxYnpedpUgcbaVxKBN1sWD unI=
3rl2q58205687c8i9kc9mv46dghcns45.com. 86400 IN NSEC3 1 1 0 - 3RL3ODP8D910939I655B97GAQU6VE1Q7 NS DS RRSIG
3rl2q58205687c8i9kc9mv46dghcns45.com. 86400 IN RRSIG NSEC3 8 2 86400 20190130060506 20190123045506 16883 com. IQCRFT2YkQzQfdBaztXfbl+Rz7aOZm5QFHRE8P1KMct4Rh0cG59UHSfS 5o6ftVehczm1ZJfBH+suvtsuHLRaQiJ2b4DScmQx/KLVcyO5HvG2irKd MBMX1cpZQgGn5z18GB7CEF2TfoQlep+W1JR1ve7kfmtbxfv7cUFDPgl+ 2LE=
```

**Comment by bmrzycki** [comment]
Thanks for the reply @aarond10 . As long as they are related to DNSSEC I'm not too worried about them. I'm actually quite impressed quad9 worked with minimal configuration changes, nice work!

**Comment by aarond10** [comment]
I'd love to support these two but it's a bit of a hassle and will take a
bit of effort. Needs a mapping from rrtype name to id, time parsing, and
base64.

On 25 Jan. 2019 1:18 pm, "Brian Rzycki" <notifications@github.com> wrote:

Thanks for the reply @aarond10 <https://github.com/aarond10> . As long as
they are related to DNSSEC I'm not too worried about them. I'm actually
quite impressed quad9 worked with minimal configuration changes, nice work!

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/aarond10/https_dns_proxy/issues/50#issuecomment-457431289>,
or mute the thread

**Comment by bmrzycki** [comment]
Hi @aarond10 , thank you for taking the time to work on this. I am unable to build the newest code on my Ubuntu-variant host. It complains about `u_char` being undefined:

```
bmr@chrx ~ $ mkdir scratch
bmr@chrx ~ $ cd scratch/
bmr@chrx ~/scratch $ git clone https://github.com/aarond10/https_dns_proxy.git
Cloning into 'https_dns_proxy'...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (9/9), done.
Receiving objects: 100% (509/509), 135.05 KiB | 0 bytes/s, done.
remote: Total 509 (delta 3), reused 5 (delta 2), pack-reused 498
Resolving deltas: 100% (339/339), done.
Checking connectivity... done.
bmr@chrx ~/scratch $ cd https_dns_proxy/
bmr@chrx ~/scratch/https_dns_proxy $ cmake .
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- clang-tidy not found.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bmr/scratch/https_dns_proxy
bmr@chrx ~/scratch/https_dns_proxy $ make -j 1
Scanning dependencies of target nxjson
[ 10%] Building C object lib/CMakeFiles/nxjson.dir/nxjson/nxjson.c.o
[ 20%] Linking C static library libnxjson.a
[ 20%] Built target nxjson
Scanning dependencies of target https_dns_proxy
[ 30%] Building C object CMakeFiles/https_dns_proxy.dir/src/https_client.c.o
[ 40%] Building C object CMakeFiles/https_dns_proxy.dir/src/options.c.o
[ 50%] Building C object CMakeFiles/https_dns_proxy.dir/src/dns_poller.c.o
[ 60%] Building C object CMakeFiles/https_dns_proxy.dir/src/main.c.o
[ 70%] Building C object CMakeFiles/https_dns_proxy.dir/src/dns_server.c.o
[ 80%] Building C object CMakeFiles/https_dns_proxy.dir/src/logging.c.o
[ 90%] Building C object CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o
In file included from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:8:0:
/usr/include/arpa/nameser.h:115:8: error: unknown type name ‘u_char’
  const u_char *_msg, *_eom;
        ^
/usr/include/arpa/nameser.h:117:8: error: unknown type name ‘u_char’
  const u_char *_sections[ns_s_max];
        ^
/usr/include/arpa/nameser.h:120:8: error: unknown type name ‘u_char’
  const u_char *_msg_ptr;
        ^
/usr/include/arpa/nameser.h:144:8: error: unknown type name ‘u_char’
  const u_char * rdata;
        ^
/usr/include/arpa/nameser.h:473:1: error: unknown type name ‘u_int’
 u_int  ns_get16 (const u_char *) __THROW;
 ^
/usr/include/arpa/nameser.h:473:24: error: unknown type name ‘u_char’
 u_int  ns_get16 (const u_char *) __THROW;
                        ^
/usr/include/arpa/nameser.h:474:1: error: unknown type name ‘u_long’
 u_long  ns_get32 (const u_char *) __THROW;
 ^
/usr/include/arpa/nameser.h:474:25: error: unknown type name ‘u_char’
 u_long  ns_get32 (const u_char *) __THROW;
                         ^
/usr/include/arpa/nameser.h:475:31: error: expected ‘)’ before ‘*’ token
 void  ns_put16 (u_int, u_char *) __THROW;
                               ^
/usr/include/arpa/nameser.h:476:32: error: expected ‘)’ before ‘*’ token
 void  ns_put32 (u_long, u_char *) __THROW;
                                ^
/usr/include/arpa/nameser.h:477:26: error: unknown type name ‘u_char’
 int  ns_initparse (const u_char *, int, ns_msg *) __THROW;
                          ^
/usr/include/arpa/nameser.h:478:23: error: unknown type name ‘u_char’
 int  ns_skiprr (const u_char *, const u_char *, ns_sect, int)
                       ^
/usr/include/arpa/nameser.h:478:39: error: unknown type name ‘u_char’
 int  ns_skiprr (const u_char *, const u_char *, ns_sect, int)
                                       ^
/usr/include/arpa/nameser.h:484:26: error: unknown type name ‘u_char’
 int  ns_sprintrrf (const u_char *, size_t, const char *,
                          ^
/usr/include/arpa/nameser.h:485:29: error: unknown type name ‘u_long’
          ns_class, ns_type, u_long, const u_char *,
                             ^
/usr/include/arpa/nameser.h:485:43: error: unknown type name ‘u_char’
          ns_class, ns_type, u_long, const u_char *,
                                           ^
/usr/include/arpa/nameser.h:488:29: error: expected ‘)’ before ‘char’
 int  ns_format_ttl (u_long, char *, size_t) __THROW;
                             ^
/usr/include/arpa/nameser.h:489:34: error: unknown type name ‘u_long’
 int  ns_parse_ttl (const char *, u_long *) __THROW;
                                  ^
/usr/include/arpa/nameser.h:491:26: error: unknown type name ‘u_char’
 int  ns_name_ntol (const u_char *, u_char *, size_t) __THROW;
                          ^
/usr/include/arpa/nameser.h:491:36: error: unknown type name ‘u_char’
 int  ns_name_ntol (const u_char *, u_char *, size_t) __THROW;
                                    ^
/usr/include/arpa/nameser.h:492:26: error: unknown type name ‘u_char’
 int  ns_name_ntop (const u_char *, char *, size_t) __THROW;
                          ^
/usr/include/arpa/nameser.h:493:34: error: unknown type name ‘u_char’
 int  ns_name_pton (const char *, u_char *, size_t) __THROW;
                                  ^
/usr/include/arpa/nameser.h:494:28: error: unknown type name ‘u_char’
 int  ns_name_unpack (const u_char *, const u_char *,
                            ^
/usr/include/arpa/nameser.h:494:44: error: unknown type name ‘u_char’
 int  ns_name_unpack (const u_char *, const u_char *,
                                            ^
/usr/include/arpa/nameser.h:495:11: error: unknown type name ‘u_char’
     const u_char *, u_char *, size_t) __THROW;
           ^
/usr/include/arpa/nameser.h:495:21: error: unknown type name ‘u_char’
     const u_char *, u_char *, size_t) __THROW;
                     ^
/usr/include/arpa/nameser.h:496:26: error: unknown type name ‘u_char’
 int  ns_name_pack (const u_char *, u_char *, int,
                          ^
/usr/include/arpa/nameser.h:496:36: error: unknown type name ‘u_char’
 int  ns_name_pack (const u_char *, u_char *, int,
                                    ^
/usr/include/arpa/nameser.h:497:16: error: unknown type name ‘u_char’
          const u_char **, const u_char **) __THROW;
                ^
/usr/include/arpa/nameser.h:497:33: error: unknown type name ‘u_char’
          const u_char **, const u_char **) __THROW;
                                 ^
/usr/include/arpa/nameser.h:498:32: error: unknown type name ‘u_char’
 int  ns_name_uncompress (const u_char *, const u_char *,
                                ^
/usr/include/arpa/nameser.h:498:48: error: unknown type name ‘u_char’
 int  ns_name_uncompress (const u_char *, const u_char *,
                                                ^
/usr/include/arpa/nameser.h:499:15: error: unknown type name ‘u_char’
         const u_char *, char *, size_t) __THROW;
               ^
/usr/include/arpa/nameser.h:500:38: error: unknown type name ‘u_char’
 int  ns_name_compress (const char *, u_char *, size_t,
                                      ^
/usr/include/arpa/nameser.h:501:13: error: unknown type name ‘u_char’
       const u_char **, const u_char **) __THROW;
             ^
/usr/include/arpa/nameser.h:501:30: error: unknown type name ‘u_char’
       const u_char **, const u_char **) __THROW;
                              ^
/usr/include/arpa/nameser.h:502:26: error: unknown type name ‘u_char’
 int  ns_name_skip (const u_char **, const u_char *) __THROW;
                          ^
/usr/include/arpa/nameser.h:502:43: error: unknown type name ‘u_char’
 int  ns_name_skip (const u_char **, const u_char *) __THROW;
                                           ^
/usr/include/arpa/nameser.h:503:31: error: unknown type name ‘u_char’
 void  ns_name_rollback (const u_char *, const u_char **,
                               ^
/usr/include/arpa/nameser.h:503:47: error: unknown type name ‘u_char’
 void  ns_name_rollback (const u_char *, const u_char **,
                                               ^
/usr/include/arpa/nameser.h:504:13: error: unknown type name ‘u_char’
       const u_char **) __THROW;
             ^
/usr/include/arpa/nameser.h:505:15: error: unknown type name ‘u_char’
 int  ns_sign (u_char *, int *, int, int, void *,
               ^
/usr/include/arpa/nameser.h:506:11: error: unknown type name ‘u_char’
     const u_char *, int, u_char *, int *, time_t) __THROW;
           ^
/usr/include/arpa/nameser.h:506:26: error: unknown type name ‘u_char’
     const u_char *, int, u_char *, int *, time_t) __THROW;
                          ^
/usr/include/arpa/nameser.h:507:16: error: unknown type name ‘u_char’
 int  ns_sign2 (u_char *, int *, int, int, void *,
                ^
/usr/include/arpa/nameser.h:508:12: error: unknown type name ‘u_char’
      const u_char *, int, u_char *, int *, time_t,
            ^
/usr/include/arpa/nameser.h:508:27: error: unknown type name ‘u_char’
      const u_char *, int, u_char *, int *, time_t,
                           ^
/usr/include/arpa/nameser.h:509:6: error: unknown type name ‘u_char’
      u_char **, u_char **) __THROW;
      ^
/usr/include/arpa/nameser.h:509:17: error: unknown type name ‘u_char’
      u_char **, u_char **) __THROW;
                 ^
/usr/include/arpa/nameser.h:510:19: error: unknown type name ‘u_char’
 int  ns_sign_tcp (u_char *, int *, int, int,
                   ^
/usr/include/arpa/nameser.h:512:20: error: unknown type name ‘u_char’
 int  ns_sign_tcp2 (u_char *, int *, int, int,
                    ^
/usr/include/arpa/nameser.h:514:10: error: unknown type name ‘u_char’
          u_char **, u_char **) __THROW;
          ^
/usr/include/arpa/nameser.h:514:21: error: unknown type name ‘u_char’
          u_char **, u_char **) __THROW;
                     ^
/usr/include/arpa/nameser.h:515:38: error: unknown type name ‘u_char’
 int  ns_sign_tcp_init (void *, const u_char *, int,
                                      ^
/usr/include/arpa/nameser.h:517:1: error: unknown type name ‘u_char’
 u_char  *ns_find_tsig (u_char *, u_char *) __THROW;
 ^
/usr/include/arpa/nameser.h:517:24: error: unknown type name ‘u_char’
 u_char  *ns_find_tsig (u_char *, u_char *) __THROW;
                        ^
/usr/include/arpa/nameser.h:517:34: error: unknown type name ‘u_char’
 u_char  *ns_find_tsig (u_char *, u_char *) __THROW;
                                  ^
/usr/include/arpa/nameser.h:518:17: error: unknown type name ‘u_char’
 int  ns_verify (u_char *, int *, void *, const u_char *, int,
                 ^
/usr/include/arpa/nameser.h:518:48: error: unknown type name ‘u_char’
 int  ns_verify (u_char *, int *, void *, const u_char *, int,
                                                ^
/usr/include/arpa/nameser.h:519:7: error: unknown type name ‘u_char’
       u_char *, int *, time_t *, int) __THROW;
       ^
/usr/include/arpa/nameser.h:520:21: error: unknown type name ‘u_char’
 int  ns_verify_tcp (u_char *, int *, ns_tcp_tsig_state *, int)
                     ^
/usr/include/arpa/nameser.h:522:40: error: unknown type name ‘u_char’
 int  ns_verify_tcp_init (void *, const u_char *, int,
                                        ^
In file included from /usr/include/arpa/nameser.h:531:0,
                 from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:8:
/usr/include/arpa/nameser_compat.h:66:11: error: duplicate member ‘rd’
  unsigned rd :1;  /*%< recursion desired */
           ^
/usr/include/arpa/nameser_compat.h:67:11: error: duplicate member ‘tc’
  unsigned tc :1;  /*%< truncated message */
           ^
/usr/include/arpa/nameser_compat.h:68:11: error: duplicate member ‘aa’
  unsigned aa :1;  /*%< authoritive answer */
           ^
/usr/include/arpa/nameser_compat.h:69:11: error: duplicate member ‘opcode’
  unsigned opcode :4; /*%< purpose of message */
           ^
/usr/include/arpa/nameser_compat.h:70:11: error: duplicate member ‘qr’
  unsigned qr :1;  /*%< response flag */
           ^
/usr/include/arpa/nameser_compat.h:72:11: error: duplicate member ‘rcode’
  unsigned rcode :4; /*%< response code */
           ^
/usr/include/arpa/nameser_compat.h:73:11: error: duplicate member ‘cd’
  unsigned cd: 1;  /*%< checking disabled by resolver */
           ^
/usr/include/arpa/nameser_compat.h:74:11: error: duplicate member ‘ad’
  unsigned ad: 1;  /*%< authentic data from named */
           ^
/usr/include/arpa/nameser_compat.h:75:11: error: duplicate member ‘unused’
  unsigned unused :1; /*%< unused bits (MBZ as of 4.9.3a3) */
           ^
/usr/include/arpa/nameser_compat.h:76:11: error: duplicate member ‘ra’
  unsigned ra :1;  /*%< recursion available */
           ^
In file included from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:11:0:
/usr/include/resolv.h:75:15: error: unknown type name ‘u_char’
         const u_char **__query,
               ^
/usr/include/resolv.h:77:9: error: unknown type name ‘u_char’
         u_char *__ans,
         ^
/usr/include/resolv.h:82:15: error: unknown type name ‘u_char’
         const u_char *__query,
               ^
/usr/include/resolv.h:84:9: error: unknown type name ‘u_char’
         u_char *__ans,
         ^
/usr/include/resolv.h:107:2: error: unknown type name ‘u_long’
  u_long options;  /* option flags - see below. */
  ^
/usr/include/resolv.h:112:2: error: unknown type name ‘u_short’
  u_short id;   /* current message id */
  ^
/usr/include/resolv.h:116:2: error: unknown type name ‘u_long’
  u_long pfcode;   /* RES_PRF_ flags - see below. */
  ^
/usr/include/resolv.h:126:2: error: unknown type name ‘res_send_qhook’
  res_send_qhook qhook;  /* query hook */
  ^
/usr/include/resolv.h:127:2: error: unknown type name ‘res_send_rhook’
  res_send_rhook rhook;  /* response hook */
  ^
/usr/include/resolv.h:130:2: error: unknown type name ‘u_int’
  u_int _flags;   /* PRIVATE: see below */
  ^
/usr/include/resolv.h:268:24: error: unknown type name ‘u_char’
 void  fp_nquery (const u_char *, int, FILE *) __THROW;
                        ^
/usr/include/resolv.h:269:23: error: unknown type name ‘u_char’
 void  fp_query (const u_char *, FILE *) __THROW;
                       ^
/usr/include/resolv.h:271:22: error: unknown type name ‘u_char’
 void  p_query (const u_char *) __THROW;
                      ^
/usr/include/resolv.h:275:54: error: unknown type name ‘u_char’
 int  res_mkquery (int, const char *, int, int, const u_char *,
                                                      ^
/usr/include/resolv.h:276:20: error: unknown type name ‘u_char’
         int, const u_char *, u_char *, int) __THROW;
                    ^
/usr/include/resolv.h:276:30: error: unknown type name ‘u_char’
         int, const u_char *, u_char *, int) __THROW;
                              ^
/usr/include/resolv.h:277:41: error: unknown type name ‘u_char’
 int  res_query (const char *, int, int, u_char *, int) __THROW;
                                         ^
/usr/include/resolv.h:279:6: error: unknown type name ‘u_char’
      u_char *, int) __THROW;
      ^
/usr/include/resolv.h:280:42: error: unknown type name ‘u_char’
 int  res_search (const char *, int, int, u_char *, int) __THROW;
                                          ^
/usr/include/resolv.h:281:22: error: unknown type name ‘u_char’
 int  res_send (const u_char *, int, u_char *, int) __THROW;
                      ^
/usr/include/resolv.h:281:37: error: unknown type name ‘u_char’
 int  res_send (const u_char *, int, u_char *, int) __THROW;
                                     ^
/usr/include/resolv.h:335:23: error: expected ‘)’ before ‘const’
 int  b64_ntop (u_char const *, size_t, char *, size_t) __THROW;
                       ^
/usr/include/resolv.h:336:30: error: unknown type name ‘u_char’
 int  b64_pton (char const *, u_char *, size_t) __THROW;
                              ^
/usr/include/resolv.h:337:37: error: unknown type name ‘u_char’
 int  loc_aton (const char *__ascii, u_char *__binary) __THROW;
                                     ^
/usr/include/resolv.h:338:30: error: unknown type name ‘u_char’
 const char * loc_ntoa (const u_char *__binary, char *__ascii) __THROW;
                              ^
/usr/include/resolv.h:339:25: error: unknown type name ‘u_char’
 int  dn_skipname (const u_char *, const u_char *) __THROW;
                         ^
/usr/include/resolv.h:339:41: error: unknown type name ‘u_char’
 int  dn_skipname (const u_char *, const u_char *) __THROW;
                                         ^
/usr/include/resolv.h:340:27: error: unknown type name ‘u_char’
 void  putlong (u_int32_t, u_char *) __THROW;
                           ^
/usr/include/resolv.h:341:28: error: unknown type name ‘u_char’
 void  putshort (u_int16_t, u_char *) __THROW;
                            ^
/usr/include/resolv.h:346:7: error: unknown type name ‘u_char’
 const u_char * p_cdnname (const u_char *, const u_char *, int, FILE *)
       ^
/usr/include/resolv.h:346:33: error: unknown type name ‘u_char’
 const u_char * p_cdnname (const u_char *, const u_char *, int, FILE *)
                                 ^
/usr/include/resolv.h:346:49: error: unknown type name ‘u_char’
 const u_char * p_cdnname (const u_char *, const u_char *, int, FILE *)
                                                 ^
/usr/include/resolv.h:348:7: error: unknown type name ‘u_char’
 const u_char * p_cdname (const u_char *, const u_char *, FILE *) __THROW;
       ^
/usr/include/resolv.h:348:32: error: unknown type name ‘u_char’
 const u_char * p_cdname (const u_char *, const u_char *, FILE *) __THROW;
                                ^
/usr/include/resolv.h:348:48: error: unknown type name ‘u_char’
 const u_char * p_cdname (const u_char *, const u_char *, FILE *) __THROW;
                                                ^
/usr/include/resolv.h:349:7: error: unknown type name ‘u_char’
 const u_char * p_fqnname (const u_char *__cp, const u_char *__msg,
       ^
/usr/include/resolv.h:349:33: error: unknown type name ‘u_char’
 const u_char * p_fqnname (const u_char *__cp, const u_char *__msg,
                                 ^
/usr/include/resolv.h:349:53: error: unknown type name ‘u_char’
 const u_char * p_fqnname (const u_char *__cp, const u_char *__msg,
                                                     ^
/usr/include/resolv.h:351:7: error: unknown type name ‘u_char’
 const u_char * p_fqname (const u_char *, const u_char *, FILE *) __THROW;
       ^
/usr/include/resolv.h:351:32: error: unknown type name ‘u_char’
 const u_char * p_fqname (const u_char *, const u_char *, FILE *) __THROW;
                                ^
/usr/include/resolv.h:351:48: error: unknown type name ‘u_char’
 const u_char * p_fqname (const u_char *, const u_char *, FILE *) __THROW;
                                                ^
/usr/include/resolv.h:352:24: error: unknown type name ‘u_long’
 const char * p_option (u_long __option) __THROW;
                        ^
/usr/include/resolv.h:355:29: error: unknown type name ‘u_char’
 int  dn_comp (const char *, u_char *, int, u_char **, u_char **)
                             ^
/usr/include/resolv.h:355:44: error: unknown type name ‘u_char’
 int  dn_comp (const char *, u_char *, int, u_char **, u_char **)
                                            ^
/usr/include/resolv.h:355:55: error: unknown type name ‘u_char’
 int  dn_comp (const char *, u_char *, int, u_char **, u_char **)
                                                       ^
/usr/include/resolv.h:357:23: error: unknown type name ‘u_char’
 int  dn_expand (const u_char *, const u_char *, const u_char *,
                       ^
/usr/include/resolv.h:357:39: error: unknown type name ‘u_char’
 int  dn_expand (const u_char *, const u_char *, const u_char *,
                                       ^
/usr/include/resolv.h:357:55: error: unknown type name ‘u_char’
 int  dn_expand (const u_char *, const u_char *, const u_char *,
                                                       ^
/usr/include/resolv.h:359:1: error: unknown type name ‘u_int’
 u_int  res_randomid (void) __THROW;
 ^
/usr/include/resolv.h:361:12: error: unknown type name ‘u_char’
      const u_char *, const u_char *) __THROW;
            ^
/usr/include/resolv.h:361:28: error: unknown type name ‘u_char’
      const u_char *, const u_char *) __THROW;
                            ^
/usr/include/resolv.h:362:30: error: unknown type name ‘u_char’
 int  res_queriesmatch (const u_char *, const u_char *,
                              ^
/usr/include/resolv.h:362:46: error: unknown type name ‘u_char’
 int  res_queriesmatch (const u_char *, const u_char *,
                                              ^
/usr/include/resolv.h:363:13: error: unknown type name ‘u_char’
       const u_char *, const u_char *) __THROW;
             ^
/usr/include/resolv.h:363:29: error: unknown type name ‘u_char’
       const u_char *, const u_char *) __THROW;
                             ^
/usr/include/resolv.h:370:43: error: unknown type name ‘u_char’
 void  res_npquery (const res_state, const u_char *, int, FILE *)
                                           ^
/usr/include/resolv.h:374:53: error: unknown type name ‘u_char’
 int  res_nquery (res_state, const char *, int, int, u_char *, int)
                                                     ^
/usr/include/resolv.h:376:54: error: unknown type name ‘u_char’
 int  res_nsearch (res_state, const char *, int, int, u_char *, int)
                                                      ^
/usr/include/resolv.h:379:12: error: unknown type name ‘u_char’
       int, u_char *, int) __THROW;
            ^
/usr/include/resolv.h:381:16: error: unknown type name ‘u_char’
          const u_char *, int, const u_char *, u_char *,
                ^
/usr/include/resolv.h:381:37: error: unknown type name ‘u_char’
          const u_char *, int, const u_char *, u_char *,
                                     ^
/usr/include/resolv.h:381:47: error: unknown type name ‘u_char’
          const u_char *, int, const u_char *, u_char *,
                                               ^
/usr/include/resolv.h:383:34: error: unknown type name ‘u_char’
 int  res_nsend (res_state, const u_char *, int, u_char *, int)
                                  ^
/usr/include/resolv.h:383:49: error: unknown type name ‘u_char’
 int  res_nsend (res_state, const u_char *, int, u_char *, int)
                                                 ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c: In function ‘str_to_rrtype’:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:229:44: error: ‘ns_t_dnskey’ undeclared (first use in this function)
   if (!strcasecmp(str, "DNSKEY")) { return ns_t_dnskey; }
                                            ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:229:44: note: each undeclared identifier is reported only once for each function it appears in
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:230:40: error: ‘ns_t_ds’ undeclared (first use in this function)
   if (!strcasecmp(str, "DS")) { return ns_t_ds; }
                                        ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:253:42: error: ‘ns_t_nsec’ undeclared (first use in this function)
   if (!strcasecmp(str, "NSEC")) { return ns_t_nsec; }
                                          ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:254:43: error: ‘ns_t_nsec3’ undeclared (first use in this function)
   if (!strcasecmp(str, "NSEC3")) { return ns_t_nsec3; }
                                           ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:261:43: error: ‘ns_t_rrsig’ undeclared (first use in this function)
   if (!strcasecmp(str, "RRSIG")) { return ns_t_rrsig; }
                                           ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:267:43: error: ‘ns_t_sshfp’ undeclared (first use in this function)
   if (!strcasecmp(str, "SSHFP")) { return ns_t_sshfp; }
                                           ^
In file included from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:8:0:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c: In function ‘json_to_rdata’:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:423:3: error: unknown type name ‘u_char’
   NS_PUT16(0xffff, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:423:3: error: ‘u_char’ undeclared (first use in this function)
   NS_PUT16(0xffff, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:423:3: error: expected expression before ‘)’ token
   NS_PUT16(0xffff, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:447:5: error: ‘t_cp’ undeclared (first use in this function)
     NS_PUT16(prio, pos);
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:447:5: error: expected expression before ‘)’ token
     NS_PUT16(prio, pos);
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:539:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // serial
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:540:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // refresh
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:541:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // retry
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:542:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // expire
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:543:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // min
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:548:5: error: expected expression before ‘)’ token
     NS_PUT16(atoi(strtok_r(data, " ", &saveptr)), pos); // prio
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:549:5: error: expected expression before ‘)’ token
     NS_PUT16(atoi(strtok_r(NULL, " ", &saveptr)), pos); // weight
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:550:5: error: expected expression before ‘)’ token
     NS_PUT16(atoi(strtok_r(NULL, " ", &saveptr)), pos); // port
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:560:8: error: ‘ns_t_rrsig’ undeclared (first use in this function)
   case ns_t_rrsig: {
        ^
In file included from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:8:0:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:563:5: error: expected expression before ‘)’ token
     NS_PUT16(str_to_rrtype(strtok_r(data, " ", &saveptr)), pos); // type
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:566:5: error: expected expression before ‘)’ token
     NS_PUT32(atoi(strtok_r(NULL, " ", &saveptr)), pos); // orig_ttl
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:567:5: error: expected expression before ‘)’ token
     NS_PUT32(parse_time(strtok_r(NULL, " ", &saveptr)), pos); // sig_expiration
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:568:5: error: expected expression before ‘)’ token
     NS_PUT32(parse_time(strtok_r(NULL, " ", &saveptr)), pos); // sig_inception
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:569:5: error: expected expression before ‘)’ token
     NS_PUT16(atoi(strtok_r(NULL, " ", &saveptr)), pos); // key_tag
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:586:8: error: ‘ns_t_nsec’ undeclared (first use in this function)
   case ns_t_nsec: {
        ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:605:8: error: ‘ns_t_nsec3’ undeclared (first use in this function)
   case ns_t_nsec3: {
        ^
In file included from /home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:8:0:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:610:5: error: expected expression before ‘)’ token
     NS_PUT16(atoi(strtok_r(NULL, " ", &saveptr)), pos); // iterations
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:641:3: error: expected expression before ‘)’ token
   NS_PUT16(r, enc_len_pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c: In function ‘json_to_dns’:
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:665:3: error: unknown type name ‘u_char’
   NS_PUT16(tx_id, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:665:3: error: ‘u_char’ undeclared (first use in this function)
   NS_PUT16(tx_id, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:665:3: error: expected expression before ‘)’ token
   NS_PUT16(tx_id, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:666:3: error: ‘t_cp’ undeclared (first use in this function)
   NS_PUT16(flags, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:666:3: error: expected expression before ‘)’ token
   NS_PUT16(flags, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:667:3: error: expected expression before ‘)’ token
   NS_PUT16(nx_json_get(json, "Question")->length, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:668:3: error: expected expression before ‘)’ token
   NS_PUT16(nx_json_get(json, "Answer")->length, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:669:3: error: expected expression before ‘)’ token
   NS_PUT16(nx_json_get(json, "Authority")->length, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:670:3: error: expected expression before ‘)’ token
   NS_PUT16(nx_json_get(json, "Additional")->length, pos);
   ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:691:5: error: expected expression before ‘)’ token
     NS_PUT16(nx_json_get(subobj, "type")->int_value, pos);
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:692:5: error: expected expression before ‘)’ token
     NS_PUT16(ns_c_in, pos);
     ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:714:9: error: expected expression before ‘)’ token
         NS_PUT16(type, pos);
         ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:715:9: error: expected expression before ‘)’ token
         NS_PUT16(ns_c_in, pos);
         ^
/home/bmr/scratch/https_dns_proxy/src/json_to_dns.c:716:9: error: expected expression before ‘)’ token
         NS_PUT32(nx_json_get(subobj, "TTL")->int_value, pos);
         ^
CMakeFiles/https_dns_proxy.dir/build.make:206: recipe for target 'CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o' failed
make[2]: *** [CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/https_dns_proxy.dir/all' failed
make[1]: *** [CMakeFiles/https_dns_proxy.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

The previous commit, SHA cac0ed65d40bda11bc466080e0717a64a967ca88 compiled without error:

```
bmr@chrx ~/scratch/https_dns_proxy $ git clean -xfd
Removing CMakeCache.txt
Removing CMakeFiles/
Removing Makefile
Removing cmake_install.cmake
Removing lib/CMakeFiles/
Removing lib/Makefile
Removing lib/cmake_install.cmake
Removing lib/libnxjson.a
bmr@chrx ~/scratch/https_dns_proxy $ git checkout HEAD~
Note: checking out 'HEAD~'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at cac0ed6... Fix a type warning over sign.
bmr@chrx ~/scratch/https_dns_proxy $ cmake .
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- clang-tidy not found.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bmr/scratch/https_dns_proxy
bmr@chrx ~/scratch/https_dns_proxy $ make
Scanning dependencies of target nxjson
[ 10%] Building C object lib/CMakeFiles/nxjson.dir/nxjson/nxjson.c.o
[ 20%] Linking C static library libnxjson.a
[ 20%] Built target nxjson
Scanning dependencies of target https_dns_proxy
[ 30%] Building C object CMakeFiles/https_dns_proxy.dir/src/https_client.c.o
[ 40%] Building C object CMakeFiles/https_dns_proxy.dir/src/options.c.o
[ 50%] Building C object CMakeFiles/https_dns_proxy.dir/src/dns_poller.c.o
[ 60%] Building C object CMakeFiles/https_dns_proxy.dir/src/main.c.o
[ 70%] Building C object CMakeFiles/https_dns_proxy.dir/src/dns_server.c.o
[ 80%] Building C object CMakeFiles/https_dns_proxy.dir/src/logging.c.o
[ 90%] Building C object CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o
[100%] Linking C executable https_dns_proxy
[100%] Built target https_dns_proxy
```

**Comment by rampageX** [comment]
Build system: Tomatoware on Router, based on uClibc, Linux Buffalo 3.10.101

I faced same build problem here, and i use the ugly hack then build successful:

1. for `unknown type name 'u_char'` etc, comment:  `#define _XOPEN_SOURCE` in json_to_dns.c;

2. for `error: 'ns_t_rrsig' undeclared` etc, **backup** and replace arpa/nameser.h with newest glibc's file resolv\arpa\nameser.h, because the old file does not include:

```
    ns_t_apl = 42,
    ns_t_ds = 43,
    ns_t_sshfp = 44,
    ns_t_rrsig = 46,
    ns_t_nsec = 47,
    ns_t_dnskey = 48,
    ns_t_nsec3 = 50,
    ns_t_tkey = 249,
```
3. I can build the new code finally, and the test:

```
./https_dns_proxy -u nobody -vvvv -p 6666 -b 9.9.9.9 -r "https://dns.quad9.net/dns-query?"
----------------------------
WARNING: Running as root without dropping privileges is NOT recommended.
----------------------------
[I] 1548729978.653040 main.c:186 Built Jan 29 2019 10:46:01.
[I] 1548729978.653220 main.c:187 System c-ares: 1.11.1-DEV
[I] 1548729978.653596 main.c:188 System libcurl: libcurl/7.63.0 OpenSSL/1.0.2q zlib/1.2.11
[I] 1548729978.671475 dns_server.c:39 Listening on 127.0.0.1:6666
[D] 1548729978.671625 logging.c:49 initializing periodic log flush timer
[D] 1548729978.671916 dns_poller.c:102 Adding DNS server '9.9.9.9' for bootstrap resolution.
[I] 1548729978.672839 main.c:252 DNS polling initialized for 'dns.quad9.net'
[D] 1548729979.160688 main.c:155 Received new IP '149.112.112.112'
[D] 1548729979.160845 dns_poller.c:56 DNS poll interval changed from 5 -> 120
[D] 1548729982.113376 main.c:116 Received request for 'tensorflow.com' id: 7db4, type 1, flags 0120
[D] 1548729982.113677 https_client.c:57 Requesting HTTP/1.1: 0

[D] 1548729983.272811 https_client.c:96 CURLINFO_EFFECTIVE_URL: https://dns.quad9.net/dns-query?name=tensorflow.com&type=1
[D] 1548729983.272946 https_client.c:169 Times: 0.000310, 0.001054, 0.889416, 0.889827, 1.158258, 1.158709
[D] 1548729983.273160 main.c:81 buflen 2239

[D] 1548729983.273369 main.c:95 Received response for id 7db4: {"Status":3,"TC":false,"RD":true,"RA":true,"AD":false,"CD":false,"Question":[{"name":"tensorflow.com.","type":1}],"Authority":[{"name":"com.","type":6,"TTL":900,"Expires":"Tue, 29 Jan 2019 03:01:23 UTC","data":"a.gtld-servers.net. nstld.verisign-grs.com. 1548729959 1800 900 604800 86400"},{"name":"com.","type":46,"TTL":900,"Expires":"Tue, 29 Jan 2019 03:01:23 UTC","data":"SOA 8 1 900 20190205024559 20190129013559 16883 com. SuR8aasWaBf4wwEpY1folPFnCr7xJDscz5UrBuz69+4PXmwxzLF16T+mRA1RqHAQ+xwFSVk9impMLCYHHVlPaD+ptr8Ic5z/pPK+H4uG7xfrg4KarOnL9nIVwaE0ZlHW81vUmXnoC3vMSIwB5fcLoonxy2p3nWdU681YFsTduaM="},{"name":"CK0POJMG874LJREF7EFN8430QVIT8BSM.com.","type":50,"TTL":10688,"Expires":"Tue, 29 Jan 2019 05:44:31 UTC","data":"1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM"},{"name":"CK0POJMG874LJREF7EFN8430QVIT8BSM.com.","type":46,"TTL":10688,"Expires":"Tue, 29 Jan 2019 05:44:31 UTC","data":"NSEC3 8 2 86400 20190204054423 20190128043423 16883 com. a6+nG0HlYQFu/smtBipliyxAEgUrNXxqIn/9FuGfEPzfC+tJ41P3ne/0fTDxnEo7AX9FZ9GYQwhf288SiBsQuemJJCqS8H9xOp6TCQa8+p9xsQY4ZewJb9mJC2tDhke7IiQsCC4kBwVtNlzavKvVfu3MhvyeghPZEx9n+DoQF50="},{"name":"EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com.","type":50,"TTL":86400,"Expires":"Wed, 30 Jan 2019 02:46:23 UTC","data":"1 1 0 - EK3C80TGFMR3PIS2U1G8SH6PQLT1I6LT NS DS RRSIG"},{"name":"EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com.","type":46,"TTL":86400,"Expires":"Wed, 30 Jan 2019 02:46:23 UTC","data":"NSEC3 8 2 86400 20190202053040 20190126042040 16883 com. ApLTCsYjMAjKVJb6jWGyhgttt5JLRkeGm+4zNRpX5mwmESNcd0a6HmIly7B3EbP8ZZt11j6r5cZyvzRvcFqYNG/fAHh+o0Znh9kJllsRfVRene14oCmSFUERQojy8derM89qW3HTqfuOsq9ql4N7eyplIwCdc7iYpHRXWPIIiHo="},{"name":"3RL2Q58205687C8I9KC9MV46DGHCNS45.com.","type":50,"TTL":11605,"Expires":"Tue, 29 Jan 2019 05:59:48 UTC","data":"1 1 0 - 3RL3ODP8D910939I655B97GAQU6VE1Q7 NS DS RRSIG"},{"name":"3RL2Q58205687C8I9KC9MV46DGHCNS45.com.","type":46,"TTL":11605,"Expires":"Tue, 29 Jan 2019 05:59:48 UTC","data":"NSEC3 8 2 86400 20190203060150 20190127045150 16883 com. noI/XKkeOOTu/Hnuh7ZHWPa+WR0qjxHuZD7R3bS5/2yZ0ojarAArA07dEh/ep5zV7X0NRLo7BOwxMNB7X9ab2NadirCqCI1IFnxwmzQAjMzu5vScWgtkkz/o0V+qEzHk7SzmzNZYL/Js4wCefFPvfPOcRZ6WYUYK1ocehRUUT1E="}],"edns_client_subnet":"2620:fe::/0"}
[W] 1548729983.274052 json_to_dns.c:363 Invalid hex byte string.
[D] 1548729983.274160 json_to_dns.c:627 Failed hex decode.
[W] 1548729983.274278 json_to_dns.c:733 Failed to encode Authority ix 2.
[W] 1548729983.274389 json_to_dns.c:363 Invalid hex byte string.
[D] 1548729983.274486 json_to_dns.c:627 Failed hex decode.
[W] 1548729983.274537 json_to_dns.c:733 Failed to encode Authority ix 4.
[W] 1548729983.274663 json_to_dns.c:363 Invalid hex byte string.
[D] 1548729983.274777 json_to_dns.c:627 Failed hex decode.
[W] 1548729983.274851 json_to_dns.c:733 Failed to encode Authority ix 6
```

```
dig  @127.0.0.1 -p 6666 tensorflow.com +dnssec
;; Warning: Message parser reports malformed message packet.

; <<>> DiG 9.11.0-P3 <<>> @127.0.0.1 -p 6666 tensorflow.com +dnssec
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 32180
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 8, ADDITIONAL: 0

;; QUESTION SECTION:
;tensorflow.com.			IN	A

;; AUTHORITY SECTION:
com.			900	IN	SOA	a.gtld-servers.net. nstld.verisign-grs.com. 1548729959 1800 900 604800 86400
com.			900	IN	RRSIG	SOA 8 1 900 19700101000000 19700101000000 16883 com. SuR8aasWaBf4wwEpY1folPFnCr7xJDscz5UrBuz69+4PXmwxzLF16T+m RA1RqHAQ+xwFSVk9impMLCYHHVlPaD+ptr8Ic5z/pPK+H4uG7xfrg4Ka rOnL9nIVwaE0ZlHW81vUmXnoC3vMSIwB5fcLoonxy2p3nWdU681YFsTd uaM=
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 10688 IN RRSIG NSEC3 8 2 86400 19700101000000 19700101000000 16883 com. a6+nG0HlYQFu/smtBipliyxAEgUrNXxqIn/9FuGfEPzfC+tJ41P3ne/0 fTDxnEo7AX9FZ9GYQwhf288SiBsQuemJJCqS8H9xOp6TCQa8+p9xsQY4 ZewJb9mJC2tDhke7IiQsCC4kBwVtNlzavKvVfu3MhvyeghPZEx9n+DoQ F50=
EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com. 86400 IN RRSIG NSEC3 8 2 86400 19700101000000 19700101000000 16883 com. ApLTCsYjMAjKVJb6jWGyhgttt5JLRkeGm+4zNRpX5mwmESNcd0a6HmIl y7B3EbP8ZZt11j6r5cZyvzRvcFqYNG/fAHh+o0Znh9kJllsRfVRene14 oCmSFUERQojy8derM89qW3HTqfuOsq9ql4N7eyplIwCdc7iYpHRXWPII iHo=
3RL2Q58205687C8I9KC9MV46DGHCNS45.com. 11605 IN RRSIG NSEC3 8 2 86400 19700101000000 19700101000000 16883 com. noI/XKkeOOTu/Hnuh7ZHWPa+WR0qjxHuZD7R3bS5/2yZ0ojarAArA07d Eh/ep5zV7X0NRLo7BOwxMNB7X9ab2NadirCqCI1IFnxwmzQAjMzu5vSc Wgtkkz/o0V+qEzHk7SzmzNZYL/Js4wCefFPvfPOcRZ6WYUYK1ocehRUU T1E=

;; Query time: 1162 msec
;; SERVER: 127.0.0.1#6666(127.0.0.1)
;; WHEN: Tue Jan 29 10:46:23 CST 2019
;; MSG SIZE  rcvd: 859
```

4. The right respone:
```
dig tensorflow.com  +dnssec @192.168.4.1 -p 5252

; <<>> DiG 9.13.1 <<>> tensorflow.com +dnssec @192.168.4.1 -p 5252
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 41612
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 8, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 512
;; QUESTION SECTION:
;tensorflow.com.			IN	A

;; AUTHORITY SECTION:
com.			899	IN	SOA	a.gtld-servers.net. nstld.verisign-grs.com. 1548726183 1800 900 604800 86400
com.			899	IN	RRSIG	SOA 8 1 900 20190205014303 20190129003303 16883 com. DztnUwQrcVIDEV9yX6l04uR1sSDxCms1NB+vA21PsUA3TQ/s/sYu+1pm X12pO9ArexlfkPaZAYIMss2V/CwfYP+cTEPKsuCyrrjOQ8ks0wQv8lil ofPBH80DeAlx8BdbIFOpOPZ0W0E2puGHSh8bYx4mr13AWSV7AfGykpDK NtA=
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 86399 IN NSEC3 1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 86399 IN RRSIG NSEC3 8 2 86400 20190204054423 20190128043423 16883 com. a6+nG0HlYQFu/smtBipliyxAEgUrNXxqIn/9FuGfEPzfC+tJ41P3ne/0 fTDxnEo7AX9FZ9GYQwhf288SiBsQuemJJCqS8H9xOp6TCQa8+p9xsQY4 ZewJb9mJC2tDhke7IiQsCC4kBwVtNlzavKvVfu3MhvyeghPZEx9n+DoQ F50=
EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com. 86399 IN NSEC3 1 1 0 - EK3C80TGFMR3PIS2U1G8SH6PQLT1I6LT NS DS RRSIG
EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com. 86399 IN RRSIG NSEC3 8 2 86400 20190202053040 20190126042040 16883 com. ApLTCsYjMAjKVJb6jWGyhgttt5JLRkeGm+4zNRpX5mwmESNcd0a6HmIl y7B3EbP8ZZt11j6r5cZyvzRvcFqYNG/fAHh+o0Znh9kJllsRfVRene14 oCmSFUERQojy8derM89qW3HTqfuOsq9ql4N7eyplIwCdc7iYpHRXWPII iHo=
3RL2Q58205687C8I9KC9MV46DGHCNS45.com. 86399 IN NSEC3 1 1 0 - 3RL3ODP8D910939I655B97GAQU6VE1Q7 NS DS RRSIG
3RL2Q58205687C8I9KC9MV46DGHCNS45.com. 86399 IN RRSIG NSEC3 8 2 86400 20190203060150 20190127045150 16883 com. noI/XKkeOOTu/Hnuh7ZHWPa+WR0qjxHuZD7R3bS5/2yZ0ojarAArA07d Eh/ep5zV7X0NRLo7BOwxMNB7X9ab2NadirCqCI1IFnxwmzQAjMzu5vSc Wgtkkz/o0V+qEzHk7SzmzNZYL/Js4wCefFPvfPOcRZ6WYUYK1ocehRUU T1E=

;; Query time: 688 msec
;; SERVER: 192.168.4.1#5252(192.168.4.1)
;; WHEN: Tue Jan 29 09:43:23 CST 2019
;; MSG SIZE  rcvd: 1006
```

4.  We got the `RRSIG NSEC3` but `NSEC3`...

**Comment by bmrzycki** [comment]
Hi @aarond10 , unfortunately I'm still seeing the same `u_char` undefined issue.  If I apply the following patch I get further, but I'm still getting failures and warnings:

```
diff --git a/src/json_to_dns.c b/src/json_to_dns.c
index 0b71fa7..80bcc48 100644
--- a/src/json_to_dns.c
+++ b/src/json_to_dns.c
@@ -1,5 +1,3 @@
-#define _XOPEN_SOURCE       /* See feature_test_macros(7) */
-
 #include <sys/select.h>
 #include <sys/types.h>
 ```

And the errors:

```
[ 30%] Building C object CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o
/home/bmr/1/https_dns_proxy/src/json_to_dns.c: In function ‘str_to_rrtype’:
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:239:44: error: ‘ns_t_dnskey’ undeclared (first use in this function)
   if (!strcasecmp(str, "DNSKEY")) { return ns_t_dnskey; }
                                            ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:239:44: note: each undeclared identifier is reported only once for each function it appears in
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:240:40: error: ‘ns_t_ds’ undeclared (first use in this function)
   if (!strcasecmp(str, "DS")) { return ns_t_ds; }
                                        ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:263:42: error: ‘ns_t_nsec’ undeclared (first use in this function)
   if (!strcasecmp(str, "NSEC")) { return ns_t_nsec; }
                                          ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:264:43: error: ‘ns_t_nsec3’ undeclared (first use in this function)
   if (!strcasecmp(str, "NSEC3")) { return ns_t_nsec3; }
                                           ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:271:43: error: ‘ns_t_rrsig’ undeclared (first use in this function)
   if (!strcasecmp(str, "RRSIG")) { return ns_t_rrsig; }
                                           ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:277:43: error: ‘ns_t_sshfp’ undeclared (first use in this function)
   if (!strcasecmp(str, "SSHFP")) { return ns_t_sshfp; }
                                           ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c: In function ‘parse_time’:
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:291:7: warning: implicit declaration of function ‘strptime’ [-Wimplicit-function-declaration]
   if (strptime(timestr, "%Y%m%d%H%M%S%Z", &tm) == NULL) {
       ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:291:48: warning: comparison between pointer and integer
   if (strptime(timestr, "%Y%m%d%H%M%S%Z", &tm) == NULL) {
                                                ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c: In function ‘json_to_rdata’:
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:570:8: error: ‘ns_t_rrsig’ undeclared (first use in this function)
   case ns_t_rrsig: {
        ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:596:8: error: ‘ns_t_nsec’ undeclared (first use in this function)
   case ns_t_nsec: {
        ^
/home/bmr/1/https_dns_proxy/src/json_to_dns.c:615:8: error: ‘ns_t_nsec3’ undeclared (first use in this function)
   case ns_t_nsec3: {
        ^
CMakeFiles/https_dns_proxy.dir/build.make:206: recipe for target 'CMakeFiles/https_dns_proxy.dir/src/json_to_dns.c.o' failed
```

I tried moving the header includes around to see if the order mattered and it didn't help. I haven't traced all of them but I suspect `<time.h>` is recursively included and by the time I define it in `json_to_dns.c` the header guards prevent it from defining `strptime()`.

It looks like I'm having the same problem as @rampageX regarding the glibc header as well...

**Comment by aarond10** [comment]
Could you try git pull and compile again with the latest code?

**Comment by rampageX** [comment]
It's working now!~

```
dig -p 5053 @127.0.0.1 tensorflow.com

; <<>> DiG 9.13.1 <<>> -p 5053 @127.0.0.1 tensorflow.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 63515
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 8, ADDITIONAL: 0

;; QUESTION SECTION:
;tensorflow.com.			IN	A

;; AUTHORITY SECTION:
com.			883	IN	SOA	a.gtld-servers.net. nstld.verisign-grs.com. 1548768677 1800 900 604800 86400
com.			883	IN	RRSIG	SOA 8 1 900 20190205133117 20190129122117 16883 com. eFkTQ0739avDB3o6bBF7nflUt9+PDcibjrz77cn0k23qIDltv4OolFi7 uTE4lip9EUNR5UObqkXMVerxOqQHT4EBHhW8H+UirLdBWGE4yTIGxLYk 3FFcyMqVUwvdnv00EOPhrLnd5tiw9izDok+Yv53B75zGpb0xQPiwdj6z RNs=
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 58351 IN NSEC3 1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 58351 IN RRSIG NSEC3 8 2 86400 20190204054423 20190128043423 16883 com. a6+nG0HlYQFu/smtBipliyxAEgUrNXxqIn/9FuGfEPzfC+tJ41P3ne/0 fTDxnEo7AX9FZ9GYQwhf288SiBsQuemJJCqS8H9xOp6TCQa8+p9xsQY4 ZewJb9mJC2tDhke7IiQsCC4kBwVtNlzavKvVfu3MhvyeghPZEx9n+DoQ F50=
EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com. 82164 IN NSEC3 1 1 0 - EK3C80TGFMR3PIS2U1G8SH6PQLT1I6LT NS DS RRSIG
EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com. 82164 IN RRSIG NSEC3 8 2 86400 20190202053040 20190126042040 16883 com. ApLTCsYjMAjKVJb6jWGyhgttt5JLRkeGm+4zNRpX5mwmESNcd0a6HmIl y7B3EbP8ZZt11j6r5cZyvzRvcFqYNG/fAHh+o0Znh9kJllsRfVRene14 oCmSFUERQojy8derM89qW3HTqfuOsq9ql4N7eyplIwCdc7iYpHRXWPII iHo=
3RL2Q58205687C8I9KC9MV46DGHCNS45.com. 59411 IN NSEC3 1 1 0 - 3RL3ODP8D910939I655B97GAQU6VE1Q7 NS DS RRSIG
3RL2Q58205687C8I9KC9MV46DGHCNS45.com. 59411 IN RRSIG NSEC3 8 2 86400 20190203060150 20190127045150 16883 com. noI/XKkeOOTu/Hnuh7ZHWPa+WR0qjxHuZD7R3bS5/2yZ0ojarAArA07d Eh/ep5zV7X0NRLo7BOwxMNB7X9ab2NadirCqCI1IFnxwmzQAjMzu5vSc Wgtkkz/o0V+qEzHk7SzmzNZYL/Js4wCefFPvfPOcRZ6WYUYK1ocehRUU T1E=

;; Query time: 94 msec
;; SERVER: 127.0.0.1#5053(127.0.0.1)
;; WHEN: Tue Jan 29 21:31:59 CST 2019
;; MSG SIZE  rcvd: 1097
```

```
./https_dns_proxy -u nobody -g nobody -vvv -b 1.1.1.1,1.0.0.1 -r "https://dns.quad9.net/dns-query?"
[I] 1548768693.478301 main.c:186 Built Jan 29 2019 21:29:45.
[I] 1548768693.478363 main.c:187 System c-ares: 1.15.0
[I] 1548768693.478507 main.c:188 System libcurl: libcurl/7.63.0 wolfSSL/3.15.7 zlib/1.2.11 c-ares/1.15.0 nghttp2/1.35.1
[I] 1548768693.479785 dns_server.c:39 Listening on 127.0.0.1:5053
[D] 1548768693.479859 logging.c:49 initializing periodic log flush timer
[D] 1548768693.479973 dns_poller.c:102 Adding DNS server '1.1.1.1' for bootstrap resolution.
[D] 1548768693.480008 dns_poller.c:102 Adding DNS server '1.0.0.1' for bootstrap resolution.
[I] 1548768693.480416 main.c:252 DNS polling initialized for 'dns.quad9.net'
[D] 1548768693.696197 main.c:155 Received new IP '149.112.112.112'
[D] 1548768693.696248 dns_poller.c:56 DNS poll interval changed from 5 -> 120
[D] 1548768701.310026 main.c:116 Received request for 'tensorflow.com' id: d6a2, type 1, flags 0120
[D] 1548768701.310546 https_client.c:57 Requesting HTTP/1.1: 0

[D] 1548768702.460997 https_client.c:96 CURLINFO_EFFECTIVE_URL: https://dns.quad9.net/dns-query?name=tensorflow.com&type=1
[D] 1548768702.461057 https_client.c:169 Times: 0.000178, 0.088982, 0.945949, 0.946175, 1.150032, 1.150259
[D] 1548768702.461251 main.c:81 buflen 2239

[D] 1548768702.461296 main.c:95 Received response for id d6a2: {"Status":3,"TC":false,"RD":true,"RA":true,"AD":false,"CD":false,"Question":[{"name":"tensorflow.com.","type":1}],"Authority":[{"name":"com.","type":6,"TTL":900,"Expires":"Tue, 29 Jan 2019 13:46:42 UTC","data":"a.gtld-servers.net. nstld.verisign-grs.com. 1548768677 1800 900 604800 86400"},{"name":"com.","type":46,"TTL":900,"Expires":"Tue, 29 Jan 2019 13:46:42 UTC","data":"SOA 8 1 900 20190205133117 20190129122117 16883 com. eFkTQ0739avDB3o6bBF7nflUt9+PDcibjrz77cn0k23qIDltv4OolFi7uTE4lip9EUNR5UObqkXMVerxOqQHT4EBHhW8H+UirLdBWGE4yTIGxLYk3FFcyMqVUwvdnv00EOPhrLnd5tiw9izDok+Yv53B75zGpb0xQPiwdj6zRNs="},{"name":"CK0POJMG874LJREF7EFN8430QVIT8BSM.com.","type":50,"TTL":58368,"Expires":"Wed, 30 Jan 2019 05:44:30 UTC","data":"1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM"},{"name":"CK0POJMG874LJREF7EFN8430QVIT8BSM.com.","type":46,"TTL":58368,"Expires":"Wed, 30 Jan 2019 05:44:30 UTC","data":"NSEC3 8 2 86400 20190204054423 20190128043423 16883 com. a6+nG0HlYQFu/smtBipliyxAEgUrNXxqIn/9FuGfEPzfC+tJ41P3ne/0fTDxnEo7AX9FZ9GYQwhf288SiBsQuemJJCqS8H9xOp6TCQa8+p9xsQY4ZewJb9mJC2tDhke7IiQsCC4kBwVtNlzavKvVfu3MhvyeghPZEx9n+DoQF50="},{"name":"EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com.","type":50,"TTL":82181,"Expires":"Wed, 30 Jan 2019 12:21:23 UTC","data":"1 1 0 - EK3C80TGFMR3PIS2U1G8SH6PQLT1I6LT NS DS RRSIG"},{"name":"EK3B4B8DSUR37CD5JUT6RULD7PS3B17O.com.","type":46,"TTL":82181,"Expires":"Wed, 30 Jan 2019 12:21:23 UTC","data":"NSEC3 8 2 86400 20190202053040 20190126042040 16883 com. ApLTCsYjMAjKVJb6jWGyhgttt5JLRkeGm+4zNRpX5mwmESNcd0a6HmIly7B3EbP8ZZt11j6r5cZyvzRvcFqYNG/fAHh+o0Znh9kJllsRfVRene14oCmSFUERQojy8derM89qW3HTqfuOsq9ql4N7eyplIwCdc7iYpHRXWPIIiHo="},{"name":"3RL2Q58205687C8I9KC9MV46DGHCNS45.com.","type":50,"TTL":59428,"Expires":"Wed, 30 Jan 2019 06:02:10 UTC","data":"1 1 0 - 3RL3ODP8D910939I655B97GAQU6VE1Q7 NS DS RRSIG"},{"name":"3RL2Q58205687C8I9KC9MV46DGHCNS45.com.","type":46,"TTL":59428,"Expires":"Wed, 30 Jan 2019 06:02:10 UTC","data":"NSEC3 8 2 86400 20190203060150 20190127045150 16883 com. noI/XKkeOOTu/Hnuh7ZHWPa+WR0qjxHuZD7R3bS5/2yZ0ojarAArA07dEh/ep5zV7X0NRLo7BOwxMNB7X9ab2NadirCqCI1IFnxwmzQAjMzu5vScWgtkkz/o0V+qEzHk7SzmzNZYL/Js4wCefFPvfPOcRZ6WYUYK1ocehRUUT1E="}],"edns_client_subnet":"2620:fe::/0"}
```

**Comment by bmrzycki** [comment]
As of 246e1065887f6dc48352a48d155ee3f851d04e3b I am compiling successfully and seeing no `[W]` messages for DNS SEC traffic. Well done @aarond10 and thank you!

**Comment by aarond10** [comment]
Thanks @bmrzycki for the analysis! Glad to hear things are working again. I'll have to revisit the time parsing again. The translation to/from presentation format is getting quite painful but the toolkit of helper functions that I'm building up should make the pain easier for future work.
